Clarify IHostingStartup delegate calls sequence (#10889)
aspnet/Docs: aspnetcore/fundamentals/host/platform-specific-configuration.md

* Contains a hosting startup class, `ServiceKeyInjection`, which implements `IHostingStartup`. `ServiceKeyInjection` adds a pair of service strings to the app's configuration using the in-memory configuration provider ([AddInMemoryCollection](/dotnet/api/microsoft.extensions.configuration.memoryconfigurationbuilderextensions.addinmemorycollection)).
* Includes a `HostingStartup` attribute that identifies the hosting startup's namespace and class.

> The `ServiceKeyInjection` class's [Configure](/dotnet/api/microsoft.aspnetcore.hosting.ihostingstartup.configure) method uses an [IWebHostBuilder](/dotnet/api/microsoft.aspnetcore.hosting.iwebhostbuilder) to add enhancements to an app. `IHostingStartup.Configure` in the hosting startup assembly is called by the runtime before `Startup.Configure` in user code. However, delegates registered by `IHostingStartup` (for example, when calling `ConfigureAppConfiguration`) run after user code. To ensure that configuration code that relies on callbacks is executed before user code, call `UseConfiguration` to apply the configuration to the app.

Reviewer: I think the really confusing thing here is that it's comparing `IHostingStartup.Configure(IWebHostBuilder)` to `Startup.Configure(IApplicationBuilder)`. The two aren't comparable. I'm not sure why the comment "which allows user code to overwrite any configuration provided by the hosting startup assembly" was ever there, Startup.Configure can't override anything on the IWebHostBuilder. Maybe this whole paragraph should be removed.

Author: > not sure why the comment ... was ever there

I can research that, but I have a feeling that it will just come back as my blowing a comment based on an engineering issue/PR comment that review didn't catch.

I think "any configuration" was supposed to mean "app configuration" (configuration-provider-based configuration), and then it looks like the remark is **wrong** anyway based on the issue discussion (delegates from `IHostingStartup` go **after** those from the app).

It will probably help me to repro what the OP said was happening. I usually write a better bit when I see it with my own :eyes:.

I'll try again on Thursday morning. :coffee:

Reviewer: The config ordering is a real issue, but that has nothing to do with Startup.Configure. If you want IHostingStartup to add config before the application's ConfigureAppConfiguration then you need to call UseConfiguration which executes immediately rather than as a delayed callback.

Only part of the file is shown. The assembly name in the example is `StartupEnhancement`.

> ## Configuration provided by the hosting startup

Reviewer: "There are two approaches to configuration depending on if you want the app's config or the hosting startup's config to take precedence.   1. ..."

> Provide configuration to the app using <xref:Microsoft.AspNetCore.Hosting.WebHostBuilder.ConfigureAppConfiguration*> to load the configuration after the app's <xref:Microsoft.AspNetCore.Hosting.WebHostBuilder.ConfigureAppConfiguration*> delegates execute. The app can't override configuration values provided by the hosting startup using this approach.

Reviewer: ```suggestion
Provide configuration to the app using <xref:Microsoft.AspNetCore.Hosting.WebHostBuilder.ConfigureAppConfiguration*> to load the configuration after the app's <xref:Microsoft.AspNetCore.Hosting.WebHostBuilder.ConfigureAppConfiguration*> delegates execute. Hosting startup configuration will take priority over the app's configuration using this approach.
```

> Provide configuration to the app using <xref:Microsoft.AspNetCore.Hosting.HostingAbstractionsWebHostBuilderExtensions.UseConfiguration*> to load the configuration before the app's <xref:Microsoft.AspNetCore.Hosting.WebHostBuilder.ConfigureAppConfiguration*> delegates execute. The app can change the configuration values provided by the hosting startup using this approach.

Reviewer: ```suggestion
Provide configuration to the app using <xref:Microsoft.AspNetCore.Hosting.HostingAbstractionsWebHostBuilderExtensions.UseConfiguration*> to load the configuration before the app's <xref:Microsoft.AspNetCore.Hosting.WebHostBuilder.ConfigureAppConfiguration*> delegates execute. The app's  configuration values will take priority over those provided by the hosting startup using this approach.
```

```csharp
public class ConfigurationInjection : IHostingStartup
{
    public void Configure(IWebHostBuilder builder)
    {
        Dictionary<string, string> dict;

        builder.ConfigureAppConfiguration(config =>
        {
            dict = new Dictionary<string, string>
            {
                {"ConfigurationKey1", 

>                     "From IHostingStartup: Can't be overridden by the app."},

Reviewer: ```suggestion
                    "From IHostingStartup: Higher priority than the app's config."},
```

            };

            config.AddInMemoryCollection(dict);
        });

        dict = new Dictionary<string, string>
        {
            {"ConfigurationKey2", 

>                 "From IHostingStartup: The app can override this value."},

Reviewer: ```suggestion
                "From IHostingStartup: Lower priority than the app's config."},
```
